### PR TITLE
Flipswitch: Implement Classes Option

### PR DESCRIPTION
**Pull request by cgack**
Fixes gh-7715

### CODE REVIEW

File: external/jquery-ui/LICENSE.txt~Widgets: Pull in new widgets from ui and remove old widgets [hunk 1]
```diff
@@ -1,44 +0,0 @@
-Copyright 2007, 2014 jQuery Foundation and other contributors,
```
Comment (arschmitz) [inline external/jquery-ui/LICENSE.txt~Widgets: Pull in new widgets from ui and remove old widgets]: while it is correct to remove this it should not be part of this PR

File: js/widgets/forms/flipswitch.backcompat.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+		classProp: "ui-flipswitch"
+	} );
+
+	$.widget( "mobile.flipswitch", $.mobile.flipswitch, $.mobile.widget.backcompat );
+
+}
+
+return $.mobile.flipswitch;
+
+} );
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.backcompat.js]: new line at EOF

File: js/widgets/forms/flipswitch.backcompat.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+		define( [
+			"jquery",
+			"../widget.backcompat",
+			"./flipswitch" ], factory );
+	} else {
+
+		// Browser globals
+		factory( jQuery );
+	}
+})( function( $ ) {
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.backcompat.js]: missing space

File: js/widgets/forms/flipswitch.js [hunk 4]
```diff
@@ -172,13 +166,12 @@ return $.widget( "mobile.flipswitch", $.extend( {
 						.is( ":selected" ) ) ? "ui-flipswitch-active" : "" ) +
 				( element.is( ":disabled" ) ? " ui-state-disabled" : "" ) +
 				( options.corners ? " ui-corner-all" : "" ) +
-				( options.mini ? " ui-mini" : "" ) )
-			.append( on, off );
+				( options.mini ? " ui-mini" : "" ) );
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: mini should only be in backcompat

File: js/widgets/forms/flipswitch.js [hunk 5]
```diff
@@ -172,13 +166,12 @@ return $.widget( "mobile.flipswitch", $.extend( {
 						.is( ":selected" ) ) ? "ui-flipswitch-active" : "" ) +
 				( element.is( ":disabled" ) ? " ui-state-disabled" : "" ) +
 				( options.corners ? " ui-corner-all" : "" ) +
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: corners option should only be in backcompat

File: js/widgets/forms/flipswitch.js [hunk 6]
```diff
@@ -154,16 +152,12 @@ return $.widget( "mobile.flipswitch", $.extend( {
...
+		this._addClass( on, "ui-flipswitch-on", "ui-button ui-shadow ui-button-inherit" );
+		on.text( onText );
+		this._addClass( off, "ui-flipswitch-off" );
+		off.text( offText );
 
-		flipswitch
-			.addClass( "ui-flipswitch ui-shadow-inset " +
-				"ui-bar-" + theme + " " +
+		this._addClass( flipswitch, "ui-flipswitch", "ui-shadow-inset ui-bar-" + theme + " " +
 				( options.wrapperClass ? options.wrapperClass : "" ) + " " +
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: wrapper class should only be in backcompat

File: js/widgets/forms/flipswitch.js [hunk 7]
```diff
@@ -154,16 +152,12 @@ return $.widget( "mobile.flipswitch", $.extend( {
 			offText = ( this.type === "input" ) ?
 				options.offText : element.find( "option" ).eq( 0 ).text();
 
-		on
-			.addClass( "ui-flipswitch-on ui-button ui-shadow ui-button-inherit" )
-			.text( onText );
-		off
-			.addClass( "ui-flipswitch-off" )
-			.text( offText );
+		this._addClass( on, "ui-flipswitch-on", "ui-button ui-shadow ui-button-inherit" );
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: `ui-shadow` is a theme class

File: js/widgets/forms/flipswitch.js [hunk 8]
```diff
@@ -154,16 +152,12 @@ return $.widget( "mobile.flipswitch", $.extend( {
...
+		on.text( onText );
+		this._addClass( off, "ui-flipswitch-off" );
+		off.text( offText );
 
-		flipswitch
-			.addClass( "ui-flipswitch ui-shadow-inset " +
-				"ui-bar-" + theme + " " +
+		this._addClass( flipswitch, "ui-flipswitch", "ui-shadow-inset ui-bar-" + theme + " " +
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: `ui-shadow-inset` is a theme class

File: js/widgets/forms/flipswitch.js [hunk 9]
```diff
@@ -251,9 +244,8 @@ return $.widget( "mobile.flipswitch", $.extend( {
 			var currentTheme = options.theme ? options.theme : "inherit",
 				newTheme = options.theme ? options.theme : "inherit";
 
-			this.widget()
-				.removeClass( "ui-bar-" + currentTheme )
-				.addClass( "ui-bar-" + newTheme );
+			this._removeClass( this.widget(), null,  "ui-bar-" + currentTheme );
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: use the saved reference instead of a method.

File: js/widgets/forms/flipswitch.js [hunk 10]
```diff
@@ -251,9 +244,8 @@ return $.widget( "mobile.flipswitch", $.extend( {
...
+			this._addClass( this.widget(), null,  "ui-bar-" + newTheme);
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: same here

Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: seeing this makes me thing we should add a `themePrefix` property to the theme extension and perhaps a themeElement prop too? 

@gabrielschulhof what are your thoughts here? Do we have many places where a sing widget applies theme classes to multiple elements where we may want to make this a hash even?

File: tests/unit/flipswitch/flipswitch_core.js [hunk 11]
```diff
@@ -3,60 +3,60 @@
...
+test( "checkbox based flipswitch is active after right swipe", function( assert ) {
 	$( "#flip-checkbox" ).trigger( "swiperight" );
-	ok( $( "#flip-checkbox" ).parent().hasClass( "ui-flipswitch-active" ), "should not be active" );
+	assert.hasClasses( $( "#flip-checkbox" ).parent(), "ui-flipswitch-active", "should not be active" );
 } );
-test( "select based flipswitch is active after right swipe", function() {
+test( "select based flipswitch is active after right swipe", function( assert ) {
 	$( "#flip-select" ).trigger( "swiperight" );
-	ok( $( "#flip-select" ).parent().hasClass( "ui-flipswitch-active" ), "should not be active" );
+	assert.hasClasses( $( "#flip-select" ).parent(), "ui-flipswitch-active", "should not be active" );
```
Comment (arschmitz) [inline tests/unit/flipswitch/flipswitch_core.js]: i don't think any of the messages in these assertions add any value we can probably just remove them and use the default messages @cgack thoughts?

Comment (cgack) [inline tests/unit/flipswitch/flipswitch_core.js]: I think that is fair. I'll do that with the next commit

**Comment by cgack** [comment]
@arschmitz I've made some more changes, ready for you review.. I did not do the themePrefix idea yet since it still is an open question above.

File: external/jquery-ui/LICENSE.txt~Widgets [hunk 12]
```diff
@@ -41,4 +41,4 @@ CC0: http://creativecommons.org/publicdomain/zero/1.0/
 All files located in the node_modules and external directories are
 externally maintained libraries used by this software which have their
 own licenses; we recommend you read them, as their terms may differ from
-the terms above.
+the terms above.
```
Comment (arschmitz) [inline external/jquery-ui/LICENSE.txt~Widgets]: why are you changing this?

**Comment by arschmitz** [comment]
need to add `flipswitch`, `flipswitch.backcompat`, and the test files your changing to `jscs:good`

File: js/widgets/forms/flipswitch.js [hunk 13]
```diff
@@ -247,13 +240,14 @@ return $.widget( "mobile.flipswitch", $.extend( {
 	},
 
 	_setOptions: function( options ) {
+		var flipswitch = this.widget();
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: no need to do this just do `this.flipswitch`

File: js/widgets/forms/flipswitch.js [hunk 14]
```diff
@@ -42,9 +42,10 @@ return $.widget( "mobile.flipswitch", $.extend( {
 		offText: "Off",
 		theme: null,
 		enhanced: false,
-		wrapperClass: null,
-		corners: true,
-		mini: false
+		classes: {
+			"ui-flipswitch": "ui-shadow-inset",
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: missing `ui-corner-all`

**Comment by arschmitz** [comment]
This is failing jshint

File: tests/unit/flipswitch/index.html [hunk 15]
```diff
@@ -15,11 +15,13 @@
 	<link rel="stylesheet" href="../../../external/qunit/qunit.css"/>
 	<link rel="stylesheet" href="../../jqm-tests.css"/>
 	<script src="../../../external/qunit/qunit.js"></script>
+	<script src="../../../external/qunit-assert-classes/qunit-assert-classes.js"></script>
 	<script>
 		$.testHelper.asyncLoad([
 			[ "widgets/page" ],
 		[
 				"widgets/forms/flipswitch",
+				"widgets/forms/flipswitch.backcompat"
```
Comment (arschmitz) [inline tests/unit/flipswitch/index.html]: This file should set `$.ui.backcompat = false` and not include the backcompat extension. Any tests that need the backcompat should be moved to a separate file that is a copy of this file except it includes backcompat extension and tests.

File: external/jquery-ui/LICENSE.txt~Widgets (hunk 12 above)
Comment (cgack) [inline external/jquery-ui/LICENSE.txt~Widgets]: @arschmitz i dunno exactly, i added it back from the raw github because it was accidentally removed in a previous commit and didn't catch this.  probably a windows line ending thing.

**Comment by cgack** [comment]
@arschmitz this should be ready for another round of review

**Comment by arschmitz** [comment]
Awesome! ill check it out shortly

File: js/widgets/forms/flipswitch.js [hunk 16]
```diff
@@ -193,7 +186,8 @@ return $.widget( "mobile.flipswitch", $.extend( {
 
 	refresh: function() {
 		var direction,
-			existingDirection = this.flipswitch.hasClass( "ui-flipswitch-active" ) ? "_right" : "_left";
+			existingDirection = this.flipswitch
+									.hasClass( "ui-flipswitch-active" ) ? "_right" : "_left";
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: Indentation looks wonkey here

File: js/widgets/forms/flipswitch.js [hunk 17]
```diff
@@ -251,9 +245,8 @@ return $.widget( "mobile.flipswitch", $.extend( {
 			var currentTheme = options.theme ? options.theme : "inherit",
 				newTheme = options.theme ? options.theme : "inherit";
 
-			this.widget()
-				.removeClass( "ui-bar-" + currentTheme )
-				.addClass( "ui-bar-" + newTheme );
+			this._removeClass( this.flipswitch, null,  "ui-bar-" + currentTheme );
+			this._addClass( this.flipswitch, null,  "ui-bar-" + newTheme );
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: ok this is not new but this makes no sense `currentTheme` and `newTheme` cant possibly be different above so you are just removing and then adding the same class again.

File: tests/unit/flipswitch/flipswitch_core.js [hunk 18]
```diff
@@ -3,60 +3,54 @@
...
+test( "checkbox based flipswitch is active after right swipe", function( assert ) {
 	$( "#flip-checkbox" ).trigger( "swiperight" );
-	ok( $( "#flip-checkbox" ).parent().hasClass( "ui-flipswitch-active" ), "should not be active" );
+	assert.hasClasses( $( "#flip-checkbox" ).parent(), "ui-flipswitch-active" );
 } );
-test( "select based flipswitch is active after right swipe", function() {
+test( "select based flipswitch is active after right swipe", function( assert ) {
 	$( "#flip-select" ).trigger( "swiperight" );
-	ok( $( "#flip-select" ).parent().hasClass( "ui-flipswitch-active" ), "should not be active" );
+	assert.hasClasses( $( "#flip-select" ).parent(), "ui-flipswitch-active" );
```
Comment (arschmitz) [inline tests/unit/flipswitch/flipswitch_core.js]: This is not new but it seems like these could be combined there is no reason for each to be its own test that i see. 

can you combine these logically in some way?

File: js/widgets/forms/flipswitch.js (hunk 17 above)
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: Sure enough just tried and this is broken changing the theme option does not remove the old class just adds the new one.

File: tests/unit/flipswitch/backcompat-tests.html [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+				"widgets/forms/flipswitch.backcompat"
+			],
+			[ "init" ],
+			[
+				"flipswitch_backcompat.js"
+			]
+		]);
+	</script>
+
+	<script src="../../swarminject.js"></script>
```
Comment (arschmitz) [inline tests/unit/flipswitch/backcompat-tests.html]: remove this we dont use test swarm

File: tests/unit/flipswitch/flipswitch_backcompat.js [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+ * mobile flipswitch unit tests
+ */
+( function( $ ) {
+
+test( "checkbox based flipswitch is mini", function( assert ) {
+	assert.hasClasses( $( "#flip-checkbox-mini" ).parent(), "ui-mini" );
+} );
+test( "select based flipswitch is mini", function( assert ) {
+	assert.hasClasses( $( "#flip-select-mini" ).parent(), "ui-mini" );
+} );
```
Comment (arschmitz) [inline tests/unit/flipswitch/flipswitch_backcompat.js]: shouldn't there be tests for wrapper class and corners here too?

File: tests/unit/flipswitch/backcompat-tests.html [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+	<script>
+		$.testHelper.asyncLoad([
+			[ "widgets/page" ],
+			[
+				"widgets/forms/flipswitch",
+				"widgets/forms/flipswitch.backcompat"
+			],
+			[ "init" ],
+			[
+				"flipswitch_backcompat.js"
```
Comment (arschmitz) [inline tests/unit/flipswitch/backcompat-tests.html]: backcompat tests should run all the normal ones too to make sure the backcompat is not breaking anything

File: tests/unit/flipswitch/index.html [hunk 22]
```diff
@@ -15,11 +15,14 @@
...
+	<script src="../../../external/qunit-assert-classes/qunit-assert-classes.js"></script>
 	<script>
+		$.mobileBackcompat = false;
+
 		$.testHelper.asyncLoad([
 			[ "widgets/page" ],
-		[
-				"widgets/forms/flipswitch",
+			[
+				"widgets/forms/flipswitch"
```
Comment (arschmitz) [inline tests/unit/flipswitch/index.html]: your not actually setting backcompat to false anywhere here

File: build/tasks/options/jscs.js [hunk 23]
```diff
@@ -14,6 +14,8 @@ module.exports = {
 			"js/widgets/forms/button.backcompat.js",
 			"js/widgets/forms/checkboxradio.js",
 			"js/widgets/forms/checkboxradio.backcompat.js",
+			"js/widgets/forms/flipswitch.js",
+			"js/widgets/forms/flipswitch.backcompat.js",
```
Comment (arschmitz) [inline build/tasks/options/jscs.js]: can you also add the test files please

File: tests/unit/flipswitch/flipswitch_core.js (hunk 18 above)
Comment (cgack) [inline tests/unit/flipswitch/flipswitch_core.js]: @arschmitz 
why do we want to lose the granularity of these unit tests? I like knowing what each passing test has tested, is there a reason that this not preferred?

Comment (arschmitz) [inline tests/unit/flipswitch/flipswitch_core.js]: What i mean is keep all the assertions just logically group them inside fewer `test()` calls.

**Comment by cgack** [comment]
@arschmitz ready for another round!

File: tests/unit/flipswitch/backcompat-tests.html (hunk 21 above)
Comment (arschmitz) [inline tests/unit/flipswitch/backcompat-tests.html]: I think there was some miss understanding here. Don't duplicate the test file just include `flipswitch_core` here.

File: tests/unit/flipswitch/flipswitch_backcompat.js [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+	deepEqual( parseInt( $( "#flip-select" ).attr( "tabindex" ) ), -1,
+		"is untabbable - tabindex is set to -1" );
+	assert.hasClasses( $( "#flip-select-mini" ).parent(), "ui-mini", "is mini" );
+	assert.lacksClasses( $( "#flip-select-corners" ).parent(), "ui-corner-all",
+		"does not get ui-corner-all when set to false" );
+	assert.hasClasses( $( "#flip-select-wrapperclass" ).parent(), "select-custom-class",
+		"has wrapper class" );
+} );
+
+} )( jQuery );
```
Comment (arschmitz) [inline tests/unit/flipswitch/flipswitch_backcompat.js]: I still don't see any tests for `wrapperClass` or `corners` here.

File: js/widgets/forms/flipswitch.js [hunk 25]
```diff
@@ -193,7 +186,8 @@ return $.widget( "mobile.flipswitch", $.extend( {
 
 	refresh: function() {
 		var direction,
-			existingDirection = this.flipswitch.hasClass( "ui-flipswitch-active" ) ? "_right" : "_left";
+			existingDirection = this.flipswitch
+										.hasClass( "ui-flipswitch-active" ) ? "_right" : "_left";
```
Comment (arschmitz) [inline js/widgets/forms/flipswitch.js]: Too much indentation here outdent 6 tabs.

**Comment by arschmitz** [comment]
Looking good just a few small things... 
Remaining items before merge:
 * [ ] Fix tab issue https://github.com/jquery/jquery-mobile/pull/8118#discussion_r31359620
 * [ ] Add tests for `wrapperClass` and `corners` https://github.com/jquery/jquery-mobile/pull/8118#discussion_r31359168
 * [ ] Fix backcompat tests https://github.com/jquery/jquery-mobile/pull/8118#discussion_r31359032
 * [ ] Open issue on API docs to update

**Comment by cgack** [comment]
@arschmitz made the suggested changes. Opening an issue now on api docs for the changes

**Comment by arschmitz** [comment]
:+1: just need to rebase to fixup all the extra commits and the message

**Comment by cgack** [comment]
landed with f8938e0da0a4e8c5f1b6668f70d57bd8e4cd9cc3
